### PR TITLE
Transifex provider implementation

### PR DESCRIPTION
**Pull request by meatballhat**
/cc @emdantrim @diegobz

Example usage in `.travis.yml`:

```
deploy:
  provider:
    transifex:
      username: someuser
      password:
        secure: ABCD1234...5678WXYZ
```

### CODE REVIEW

File: lib/dpl/provider/transifex.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+      private
+
+      def write_transifexrc
+        File.open(File.expand_path('~/.transifexrc'), 'w') do |f|
+          f.puts <<-EOF.gsub(/^ {12}/, '')
+            [#{options[:hostname] || DEFAULT_HOSTNAME}]
+            hostname = #{options[:hostname] || DEFAULT_HOSTNAME}
+            username = #{options[:username]}
+            password = #{options[:password]}
+            token = #{options[:token]}
```
Comment (BanzaiMan) [inline lib/dpl/provider/transifex.rb]: I find `#gsub` along with the here-doc a bit misplaced. There are sevral ways to get rid of `#gsub`:

1. left align the file content
1. use ERb
1. use `DATA` after `__END__`
1. use a long string with inline `\n` (not my favorite)
1. use `#lstrip` (probably my least favorite)

File: lib/dpl/provider/transifex.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
+module DPL
+  class Provider
+    class Transifex < Provider
+      experimental 'Transifex'
+
+      DEFAULT_CLIENT_VERSION = '>=0.11'
+      DEFAULT_HOSTNAME = 'https://www.transifex.com'
+
+      def install_deploy_dependencies
+        self.class.pip 'transifex', 'transifex', options[:cli_version] || DEFAULT_CLIENT_VERSION
```
Comment (BanzaiMan) [inline lib/dpl/provider/transifex.rb]: Can we put this `||` outside of the `#pip` call? The operator precedence is not clear.

File: spec/provider/transifex_spec.rb [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+  describe '#source_push' do
+    example 'performs a tx push' do
+      expect(provider.context).to(
+        receive(:shell).with('tx push --source --no-interactive', retry: true)
+      )
+      provider.send(:source_push)
+    end
+  end
+
+  describe '#write_transifexrc' do
```
Comment (BanzaiMan) [inline spec/provider/transifex_spec.rb]: If `#write_transifexrc` is something we should test, then shouldn't it be public? Reaching in to test it with `#send` seems rather bad.

File: lib/dpl/provider/transifex.rb (hunk 1 above)
Comment (BanzaiMan) [inline lib/dpl/provider/transifex.rb]: Also, is transifex sensitive to leading white spaces? If it's not, maybe we can just drop `#gsub`.

Comment (meatballhat) [inline lib/dpl/provider/transifex.rb]: I honestly don't have an issue with stripping leading whitespace to preserve indentation continuity, nor do I understand advantages in the alternatives you've listed, but then this is a fairly subjective topic `¯\_(ツ)_/¯`.  The transifex client uses the Python stdlib config parser, afaik.  Which option would you like used?

File: lib/dpl/provider/transifex.rb (hunk 2 above)
Comment (meatballhat) [inline lib/dpl/provider/transifex.rb]: Yep :+1:

File: spec/provider/transifex_spec.rb (hunk 3 above)
Comment (meatballhat) [inline spec/provider/transifex_spec.rb]: Works for me

File: lib/dpl/provider/transifex.rb (hunk 1 above)
Comment (meatballhat) [inline lib/dpl/provider/transifex.rb]: Maybe we can instead do as [the guide says](https://github.com/bbatsov/ruby-style-guide/blob/master/README.md#heredocs)?

**Comment by BanzaiMan** [comment]
Needs documentation.

**Comment by BanzaiMan** [comment]
@meatballhat Docs, plz!

**Comment by tooomm** [comment]
Is there a documentation in the meantime??

File: lib/dpl/provider/transifex.rb [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+            "hostname = #{options[:hostname] || DEFAULT_HOSTNAME}",
+            "username = #{options[:username]}",
+            "password = #{options[:password]}",
+            "token = #{options[:token]}",
+          ].join("\n")
+        end
+      end
+
+      def source_push
+        context.shell 'tx push --source --no-interactive', retry: true
```
Comment (nextgens) [inline lib/dpl/provider/transifex.rb]: /home/travis/.rvm/gems/ruby-1.9.3-p551/gems/dpl-1.8.31/lib/dpl/cli.rb:53:in `shell': wrong number of arguments (2 for 1) (ArgumentError)
from /home/travis/.rvm/gems/ruby-1.9.3-p551/gems/dpl-1.8.31/lib/dpl/provider/transifex.rb:41:in `source_push'

https://travis-ci.org/freenet/fred/jobs/213110817

Comment (benjaoming) [inline lib/dpl/provider/transifex.rb]: Try using `cli_version` option and pin it at a lower version?

Comment (benjaoming) [inline lib/dpl/provider/transifex.rb]: Having the same issue.

Looking here: https://github.com/travis-ci/dpl/blob/master/lib/dpl/provider/pypi.rb#L96

Would it be that the `, retry: true` part is an argument that's no longer accepted?

Comment (benjaoming) [inline lib/dpl/provider/transifex.rb]: Seems like `shell` only accepts one argument: https://github.com/travis-ci/dpl/blob/master/lib/dpl/cli.rb#L53

Comment (benjaoming) [inline lib/dpl/provider/transifex.rb]:
```
/home/travis/.rvm/gems/ruby-2.2.5/gems/dpl-1.8.31/lib/dpl/cli.rb:53:in `shell': wrong number of arguments (2 for 1) (ArgumentError)
	from /home/travis/.rvm/gems/ruby-2.2.5/gems/dpl-1.8.31/lib/dpl/provider/transifex.rb:41:in `source_push'
	from /home/travis/.rvm/gems/ruby-2.2.5/gems/dpl-1.8.31/lib/dpl/provider/transifex.rb:25:in `push_app'
	from /home/travis/.rvm/gems/ruby-2.2.5/gems/dpl-1.8.31/lib/dpl/provider.rb:146:in `block in deploy'
	from /home/travis/.rvm/gems/ruby-2.2.5/gems/dpl-1.8.31/lib/dpl/cli.rb:41:in `fold'
	from /home/travis/.rvm/gems/ruby-2.2.5/gems/dpl-1.8.31/lib/dpl/provider.rb:146:in `deploy'
	from /home/travis/.rvm/gems/ruby-2.2.5/gems/dpl-1.8.31/lib/dpl/cli.rb:32:in `run'
	from /home/travis/.rvm/gems/ruby-2.2.5/gems/dpl-1.8.31/lib/dpl/cli.rb:7:in `run'
	from /home/travis/.rvm/gems/ruby-2.2.5/gems/dpl-1.8.31/bin/dpl:5:in `<top (required)>'
	from /home/travis/.rvm/gems/ruby-2.2.5/bin/dpl:23:in `load'
	from /home/travis/.rvm/gems/ruby-2.2.5/bin/dpl:23:in `<main>'
```

**Pull request merged**